src/omv/modules/py_tf.c

```diff
@@ -265,41 +265,79 @@ STATIC py_tf_model_obj_t *py_tf_load_alloc(mp_obj_t path_obj) {
+    PY_TF_SCALE_N1_1,
+    PY_TF_SCALE_N128_127,
+    PY_TF_SCALE_0_255,
+    PY_TF_SCALE_0_1_NORMALIZED,
+    PY_TF_SCALE_N1_1_NORMALIZED,
+    PY_TF_SCALE_N128_127_NORMALIZED,
+    PY_TF_SCALE_0_255_NORMALIZED
+} py_tf_input_scale_t;
+
 typedef struct py_tf_input_callback_data {
     image_t *img;
     rectangle_t *roi;
+    py_tf_input_scale_t input_scale;
 } py_tf_input_callback_data_t;
 
 STATIC void py_tf_input_callback(void *callback_data,
                                  void *model_input,
                                  libtf_parameters_t *params) {
     py_tf_input_callback_data_t *arg = (py_tf_input_callback_data_t *) callback_data;
 
-    // Disable checking input scaling and zero-point. Nets can be all over the place on the input
-    // scaling and zero-point but still work with the code below.
-
-    // if (params->input_datatype == LIBTF_DATATYPE_UINT8) {
-    //     if (fast_roundf(params->input_scale * GRAYSCALE_RANGE) != 1) {
-    //         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Expected model input scale to be 1/255!"));
-    //     }
-
-    //     if (params->input_zero_point != 0) {
-    //         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Expected model input zero point to be 0!"));
-    //     }
-    // }
-
-    // if (params->input_datatype == LIBTF_DATATYPE_INT8) {
-    //     if (fast_roundf(params->input_scale * GRAYSCALE_RANGE) != 1) {
-    //         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Expected model input scale to be 1/255!"));
-    //     }
+    int shift = (params->input_datatype == LIBTF_DATATYPE_INT8) ? GRAYSCALE_MID : 0;
+    float fscale = 1.0f, fadd = 0.0f;
 
-    //     if (params->input_zero_point != -GRAYSCALE_MID) {
-    //         mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Expected model input zero point to be -128!"));
-    //     }
-    // }
+    switch (arg->input_scale) {
+        case PY_TF_SCALE_0_1:
+        case PY_TF_SCALE_0_1_NORMALIZED:
+            fscale = 1.0f / 255.0f;
+            break;
+        case PY_TF_SCALE_N1_1:
+        case PY_TF_SCALE_N1_1_NORMALIZED:
+            fscale = 2.0f / 255.0f;
+            fadd = -1.0f;
+            break;
+        case PY_TF_SCALE_N128_127:
+        case PY_TF_SCALE_N128_127_NORMALIZED:
+            fscale = 255.0f / 127.0f;
+            fadd = -128.0f;
+            break;
+        case PY_TF_SCALE_0_255:
+        case PY_TF_SCALE_0_255_NORMALIZED:
+        default:
+            break;
+    }
 
-    int shift = (params->input_datatype == LIBTF_DATATYPE_INT8) ? GRAYSCALE_MID : 0;
-    float fscale = 1.0f / GRAYSCALE_RANGE;
+    float fscale_r = fscale, fadd_r = fadd;
+    float fscale_g = fscale, fadd_g = fadd;
+    float fscale_b = fscale, fadd_b = fadd;
+
+    // ImageNet Mean and StdDev Normalization.
+    //
+    // To normalize the input image we need to subtract the mean and divide by the standard deviation.
+    // We can do this by applying the normalization to fscale and fadd outside the loop.
```

Comment:
It's based on the ImageNet Mean/Std. These are known values.

> Also I'm not sure that scaling the mean and std first and then using them to normalize the image is the same as normalizing the image and then scaling it. Any proof that they are ?

I was following what Jan mentioned we needed to support:

{ label: 'Pixels ranging 0..1 (not normalized)', value: '0..1' },
{ label: 'Pixels ranging -1..1 (not normalized)', value: '-1..1' },
{ label: 'Pixels ranging -128..127 (not normalized)', value: '-128..127' },
{ label: 'Pixels ranging 0..255 (not normalized)', value: '0..255' },
{ label: 'PyTorch (scale to 0..1, then normalize using ImageNet mean/std)', value: 'torch' },

So, scale first and then normalize. The normalization operation is to subtract the mean and then divide by the stdev. I then worked out the math to avoid having to do two multiplies/adds in the loop and got it down to just one mutliply/add by factoring this back in.

E.g.

(((value * scale) + offset) - mean) / stdiv

Offset and mean and be combined.

(((value * scale) + (offset-mean)) / stdiv

Then stdev can be distributed

((value * scale)/stdiv) + ((offset-mean)/stdiv)

So fmul becomes (scale/fdiv) and fadd becomes ((offset-mean)/stdiv)

Reply:
> It's based on the ImageNet Mean/Std. These are known values.

If the network is trained on any other dataset, the mean and std will be different than those hard-coded values from image net. Imagine a network trained on bright/dark images. The mean and standard deviation should be args just like the scaling.

Second, about the optimization above it's just not producing the same values, unless I messed a step ?

```Python
import numpy as np

# Set the random seed for reproducibility
np.random.seed(42)

# Step 1: Generate a 4x4 random ndarray with values from 0 to 255
array = np.random.randint(0, 256, size=(4, 4))
print("Original array:\n", array)

# Step 2: Calculate the mean and standard deviation
m = np.mean(array)
s = np.std(array)
print("\nMean:", m)
print("Standard Deviation:", s)

# Step 3: Standardize the array
standardized_array = (array - m) / s
print("\nStandardized array:\n", standardized_array)

# Step 4: Scale the array from 0 to 1
min_val = np.min(standardized_array)
max_val = np.max(standardized_array)
scaled_array = (standardized_array - min_val) / (max_val - min_val)
print("\nScaled array:\n", scaled_array)

# Mean and STD were calculated on an array of 0->255
m /= 255
s /= 255

# Pretty much copied from the C code
fadd = 0.0
fscale = 1.0

# PY_TF_SCALE_0_1_NORMALIZED
fscale = 1.0 / 255.0

# Grayscale
fadd = (fadd - m) / s
fscale /= s
scaled_array = array * fscale + fadd
print("\nScaled array:\n", scaled_array)

```

Reply:
Let me check.

Reply:
Yeah, the scale you are using is different: scaled_array = (standardized_array - min_val) / (max_val - min_val)

So, that wouldn't necessarily be 1/255. My code is taking advantage of the fact that the input range is fixed and known to be 0-255. It doesn't try to maximize the dynamic range though because it doesn't need to to. The net just expects the floats to be within a certain range.

> If the network is trained on any other dataset, the mean and std will be different than those hard-coded values from image net. Imagine a network trained on bright/dark images. The mean and standard deviation should be args just like the scaling.

True, this specifically has to do with the with models that started from transfer learning off of something initially trained on ImageNet. 

Okay, let me allow you to customize this.

Reply:
Done

Reply:
> Yeah, the scale you are using is different: scaled_array = (standardized_array - min_val) / (max_val - min_val)
> 
> So, that wouldn't necessarily be 1/255. My code is taking advantage of the fact that the input range is fixed and known to be 0-255. It doesn't try to maximize the dynamic range though because it doesn't need to to. The net just expects the floats to be within a certain range.

Even if I use 1/255 I get the same output more or less:
```
Scaled array:
 [[0.4        0.70196078 0.36078431 0.05490196]
 [0.41568627 0.27843137 0.7372549  0.07843137]
 [0.4        0.4745098  0.82352941 0.83921569]
 [0.29019608 0.79215686 0.34117647 0.45490196]]
```

On the other hand, the C code doesn't seem to do any scaling, unless I missed a step:
```
Scaled array:
 [[-0.27098482  0.98410276 -0.4339832  -1.70537062]
 [-0.20578546 -0.77627982  1.13080131 -1.60757159]
 [-0.27098482  0.03871212  1.48939776  1.55459711]
 [-0.7273803   1.35899905 -0.5154824  -0.04278708]]
```

> True, this specifically has to do with the with models that started from transfer learning off of something initially trained on ImageNet.

Imagenet weights are very popular so they should be the default, but if you train something from scratch on a custom dataset, it should allow you to provide the mean and standard deviation of the input.

Reply:
The script did the order of operations wrong.

'PyTorch (scale to 0..1, then normalize using ImageNet mean/std)',

Note here that the normalized/standardized image is not from 0->1. The input just needs to be in that range since the ImageNet mean/stdev are like mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225]. So, things won't work unless the input values to this are in the range of 0->1.

```
import numpy as np

# Set the random seed for reproducibility
np.random.seed(42)

# Step 1: Generate a 4x4 random ndarray with values from 0 to 255
array = np.random.randint(0, 256, size=(4, 4))
print("Original array:\n", array)

# Step 2: Calculate the mean and standard deviation
m = np.mean(array)
s = np.std(array)
print("\nMean:", m)
print("Standard Deviation:", s)

# Step 3: Scale the array from 0 to 1
scaled_array = array / 255.0
print("\nScaled array:\n", scaled_array)

# Mean and STD were calculated on an array of 0->255
m /= 255
s /= 255

scaled_array_2 = (scaled_array - m) / s
print("\nScaled->Standarized array:\n", scaled_array_2)

# Pretty much copied from the C code
fadd = 0.0
fscale = 1.0

# PY_TF_SCALE_0_1_NORMALIZED
fscale = 1.0 / 255.0

# Grayscale
fadd = (fadd - m) / s
fscale /= s
scaled_array = array * fscale + fadd
print("\nC Code Scaled->Standarized array:\n", scaled_array)
```

```
Original array:
 [[102 179  92  14]
 [106  71 188  20]
 [102 121 210 214]
 [ 74 202  87 116]]

Mean: 118.625
Standard Deviation: 61.350300529011264

Standardized array:
 [[-0.27098482  0.98410276 -0.4339832  -1.70537062]
 [-0.20578546 -0.77627982  1.13080131 -1.60757159]
 [-0.27098482  0.03871212  1.48939776  1.55459711]
 [-0.7273803   1.35899905 -0.5154824  -0.04278708]]

Scaled array:
 [[0.44  0.825 0.39  0.   ]
 [0.46  0.285 0.87  0.03 ]
 [0.44  0.535 0.98  1.   ]
 [0.3   0.94  0.365 0.51 ]]

Scaled array:
 [[-0.27098482  0.98410276 -0.4339832  -1.70537062]
 [-0.20578546 -0.77627982  1.13080131 -1.60757159]
 [-0.27098482  0.03871212  1.48939776  1.55459711]
 [-0.7273803   1.35899905 -0.5154824  -0.04278708]]
```

Reply:
> The script did the order of operations wrong.

Not necessarily, if the model expects the input to be 0->1 then the script's order is the right order. That's what matters most, the expected input values ranges. Things get lost in chats maybe you want to double check this. For now we can just merge it as it's not that popular anyway.